File: step_13_binary_trees/13.2_medium/IB100_populate_next_right_pointers/Solution.java
import java.util.*;

/**
 * Populate Next Right Pointers (LeetCode #116)
 *
 * Connect each node to its next right node at the same level in a perfect binary tree.
 */
public class Solution {

    static class Node {
        int val;
        Node left, right, next;
        Node(int val) {
            this.val = val;
            this.next = null;
        }
    }

    // ============================================================
    // APPROACH 1: BRUTE FORCE -- BFS with Queue
    // Time: O(N)  |  Space: O(N)
    // ============================================================
    public static Node bruteForce(Node root) {
        if (root == null) return null;

        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                Node node = queue.poll();
                // Connect to next node in the level (not the last one)
                node.next = (i < size - 1) ? queue.peek() : null;

                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
        }
        return root;
    }

    // ============================================================
    // APPROACH 2: OPTIMAL -- Recursive DFS (pre-order)
    // Time: O(N)  |  Space: O(log N) recursion stack
    // ============================================================
    public static Node optimal(Node root) {
        if (root == null) return null;
        connectDFS(root);
        return root;
    }

    private static void connectDFS(Node node) {
        if (node == null || node.left == null) return; // leaf or null

        // Connect left child to right child (same parent)
        node.left.next = node.right;

        // Connect right child to next subtree's left child (cross-link)
        if (node.next != null) {
            node.right.next = node.next.left;
        }

        connectDFS(node.left);
        connectDFS(node.right);
    }

    // ============================================================
    // APPROACH 3: BEST -- O(1) Space using established next pointers
    // Time: O(N)  |  Space: O(1)
    // ============================================================
    public static Node best(Node root) {
        if (root == null) return null;

        Node leftmost = root;

        // While there is a next level to connect
        while (leftmost.left != null) {
            // Walk across the current level using next pointers
            Node curr = leftmost;
            while (curr != null) {
                // Connection 1: left child -> right child (same parent)
                curr.left.next = curr.right;

                // Connection 2: right child -> next node's left child (cross-link)
                if (curr.next != null) {
                    curr.right.next = curr.next.left;
                }

                curr = curr.next; // move to next node in current level
            }
            leftmost = leftmost.left; // go to next level
        }

        return root;
    }

    // ============================================================
    // HELPER: Print next pointers level by level
    // ============================================================
    private static void printNextPointers(Node root) {
        Node leftmost = root;
        while (leftmost != null) {
            Node curr = leftmost;
            StringBuilder sb = new StringBuilder();
            while (curr != null) {
                sb.append(curr.val);
                sb.append(" -> ");
                curr = curr.next;
            }
            sb.append("null");
            System.out.println(sb.toString());
            leftmost = leftmost.left;
        }
    }

    // ============================================================
    // MAIN
    // ============================================================
    public static void main(String[] args) {
        //      1
        //     / \
        //    2   3
        //   / \ / \
        //  4  5 6  7

        // Test brute force
        Node root1 = buildTree();
        bruteForce(root1);
        System.out.println("=== Brute Force ===");
        printNextPointers(root1);

        // Test optimal
        Node root2 = buildTree();
        optimal(root2);
        System.out.println("=== Optimal ===");
        printNextPointers(root2);

        // Test best
        Node root3 = buildTree();
        best(root3);
        System.out.println("=== Best ===");
        printNextPointers(root3);
    }

    private static Node buildTree() {
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.left.left = new Node(4);
        root.left.right = new Node(5);
        root.right.left = new Node(6);
        root.right.right = new Node(7);
        return root;
    }
}

File: step_13_binary_trees/13.2_medium/IB100_populate_next_right_pointers/solution.py
"""
Populate Next Right Pointers (LeetCode #116)

Connect each node to its next right node at the same level in a perfect binary tree.
"""
from typing import Optional
from collections import deque


class Node:
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


# ============================================================
# APPROACH 1: BRUTE FORCE -- BFS with Queue
# Time: O(N)  |  Space: O(N)
# ============================================================
def brute_force(root: Optional[Node]) -> Optional[Node]:
    if not root:
        return None

    queue = deque([root])

    while queue:
        size = len(queue)
        for i in range(size):
            node = queue.popleft()
            # Connect to next node in level (not the last one)
            node.next = queue[0] if i < size - 1 else None

            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

    return root


# ============================================================
# APPROACH 2: OPTIMAL -- Recursive DFS (pre-order)
# Time: O(N)  |  Space: O(log N) recursion stack
# ============================================================
def optimal(root: Optional[Node]) -> Optional[Node]:
    if not root:
        return None

    def connect(node):
        if not node or not node.left:
            return
        # Same parent: left -> right
        node.left.next = node.right
        # Cross-link: right -> next subtree's left
        if node.next:
            node.right.next = node.next.left
        connect(node.left)
        connect(node.right)

    connect(root)
    return root


# ============================================================
# APPROACH 3: BEST -- O(1) Space using established next pointers
# Time: O(N)  |  Space: O(1)
# ============================================================
def best(root: Optional[Node]) -> Optional[Node]:
    if not root:
        return None

    leftmost = root

    while leftmost.left:
        curr = leftmost
        while curr:
            # Connection 1: left child -> right child
            curr.left.next = curr.right
            # Connection 2: right child -> next node's left child
            if curr.next:
                curr.right.next = curr.next.left
            curr = curr.next
        leftmost = leftmost.left

    return root


# ============================================================
# HELPER: Print next pointers level by level
# ============================================================
def print_next_pointers(root):
    leftmost = root
    while leftmost:
        curr = leftmost
        parts = []
        while curr:
            parts.append(str(curr.val))
            curr = curr.next
        parts.append("null")
        print(" -> ".join(parts))
        leftmost = leftmost.left


# ============================================================
# MAIN
# ============================================================
if __name__ == "__main__":
    def build_tree():
        root = Node(1)
        root.left = Node(2)
        root.right = Node(3)
        root.left.left = Node(4)
        root.left.right = Node(5)
        root.right.left = Node(6)
        root.right.right = Node(7)
        return root

    print("=== Brute Force ===")
    r1 = build_tree()
    brute_force(r1)
    print_next_pointers(r1)

    print("=== Optimal ===")
    r2 = build_tree()
    optimal(r2)
    print_next_pointers(r2)

    print("=== Best ===")
    r3 = build_tree()
    best(r3)
    print_next_pointers(r3)
    # Expected for each:
    # 1 -> null
    # 2 -> 3 -> null
    # 4 -> 5 -> 6 -> 7 -> null
